R/xgx_stat_ci.R

```diff
@@ -99,87 +99,150 @@
+        lay$aes_params$size <- 1
+        lay$geom$geom_params$fatten <- 2
+      }
     }
-    
 
-    ret[[paste0("geom_", igeom)]] <- temp
+    lays[[paste0("geom_", igeom)]] <- lay  
   }
-
-  return(ret)
+  
+  return(lays)
 }
+
+
+#' Stat ggproto object for creating ggplot layers of binned confidence intervals
+#'
+#' 
+#' \code{StatCI} returns a ggproto object for plotting mean +/- confidence bins
+#' 
+#'
+#' @return ggplot2 ggproto object
+#'
+#' @export
+StatCI <- ggplot2::ggproto("StatCI", ggplot2::Stat,
+                  
+    required_aes = c("x","y"),
+    
+    compute_group = function(data, scales, conf_level, distribution, bins) {
+    
+      if (distribution == "ordinal") {
+        stat_data <- data
+      }
+      else {
+        stat_data <- data %>% mutate(quantile_index = dplyr::ntile(data$x, bins)) %>%
+                              group_by(quantile_index) %>%
+                              summarize(x = median(x),
+                                        ymin = xgx_conf_int(y, conf_level, distribution)$ymin,
+                                        ymax = xgx_conf_int(y, conf_level, distribution)$ymax,
+                                        y = xgx_conf_int(y, conf_level, distribution)$y) %>%
+                              ungroup()
+      }
+
+      return(stat_data)
+    },
+    
+    setup_data = function(data, params){
+      
+      # If the data is ordinal, calculate percentages
+      #   for each category across each bin
+      if (params$distribution == "ordinal"){
+        
+        # Get median x value for each bin
+        median_x <- data %>% mutate(quantile_index = dplyr::ntile(data$x, params$bins)) %>%
+                              group_by(quantile_index) %>%
+                              summarize(x = median(x))
+        
+        # Get the number of each category in each bin 
+        counts <- data %>% mutate(quantile_index = dplyr::ntile(data$x, params$bins)) %>%
+                              group_by(quantile_index, y) %>%
+                              summarize(count = length(y),
+                                        colour = unique(colour),
+                                        fill = unique(fill),
+                                        PANEL = unique(PANEL),
+                                        group = unique(group))
+
+        # Combine the x and y data
+        data <- merge(median_x, counts, by = "quantile_index", all = TRUE)
+
+        # Now calculate the confidence intervals for the multinomial data
+        data <- data %>% group_by(quantile_index) %>%
+                  mutate(x = median(x),
+                         y=as.data.frame(DescTools::MultinomCI(count, params$conf_level))$est,
+                         ymin=as.data.frame(DescTools::MultinomCI(count, params$conf_level))$lwr.ci,
+                         ymax=as.data.frame(DescTools::MultinomCI(count, params$conf_level))$upr.ci) %>%
+                  ungroup() %>% group_by(group)
```

Comment:
can you move these definitions for y, ymin, ymax to xgx_stat_ci

Reply:
y=as.data.frame(DescTools::MultinomCI(count, params$conf_level))$est,
                         ymin=as.data.frame(DescTools::MultinomCI(count, params$conf_level))$lwr.ci,
                         ymax=as.data.frame(DescTools::MultinomCI(count, params$conf_level))$upr.ci

Reply:
actually, should they go into xgx_conf_int?